### PR TITLE
[L0] Allocate event pools efficiently in multi-device scenarios

### PR DESCRIPTION
**Pull request by raiyanla**

### CODE REVIEW

**Comment by raiyanla** [comment]
CI tests should all be fixed now.

**Comment by aarongreig** [comment]
as we will be looking to merge this quickly make sure you create a (draft) PR in https://github.com/intel/llvm to pull in these changes and get them tested on the CI there ASAP @raiyanla

**Comment by raiyanla** [comment]
> as we will be looking to merge this quickly make sure you create a (draft) PR in https://github.com/intel/llvm to pull in these changes and get them tested on the CI there ASAP @raiyanla

Sure, yes will create that now, to confirm if everything is okay on that CI.

File: source/adapters/level_zero/context.hpp [hunk 1]
```diff
@@ -142,6 +142,9 @@ struct ur_context_handle_t_ : _ur_object {
   //
   // Cache of event pools to which host-visible events are added to.
   std::vector<std::list<ze_event_pool_handle_t>> ZeEventPoolCache{4};
+  std::vector<std::unordered_map<ze_device_handle_t,
+                                 std::list<ze_event_pool_handle_t> *>>
+      ZeEventPoolCacheDeviceMap{4};
```
Comment (nrspruit) [inline source/adapters/level_zero/context.hpp:147]: why 4 here as the default value? I know we have that above, I am just curious if there is a particular reason to init with 4.

Comment (raiyanla) [reply source/adapters/level_zero/context.hpp:147]: @nrspruit : It's to have a different event pool cache for each combination of scope (Host visibility) and profiling mode, as these two traits are used when originally creating an Event pool in the L0 URT.

**Comment by raiyanla** [comment]
Seeing some SYCL E2E CI tests failing in the LLVM draft PR (https://github.com/intel/llvm/pull/12666), will be attempting to resolve those ASAP.

**Review by nrspruit — APPROVED** [review]
LGTM so far after the last change, I have updated the intel/llvm patch to see if there any other issues we have missed.

**Comment by nrspruit** [comment]
Hello @oneapi-src/unified-runtime-maintain , please merge this change when available.

**Review by EwanC — APPROVED** [review]
`command_buffer.cpp` changes LGTM

**Pull request merged**